packages/task/src/browser/task-service.ts

```diff
@@ -0,0 +1,173 @@
+    }
+
+    /** Returns an array of running tasks 'TaskInfo' objects */
+    public getRunningTasks(): Promise<TaskInfo[]> {
+        return this.taskServer.getTasks(this.getContext());
+    }
+
+    /** runs a task, by task configuration label */
+    public async run(taskName: string): Promise<void> {
+        let taskInfo: TaskInfo;
+        const task = this.taskConfigurations.getTask(taskName);
+        if (!task) {
+            this.logger.error(`Can't get task launch configuration for label: ${taskName}`);
+            return;
+        }
+
+        try {
+            taskInfo = await this.taskServer.run(this.massageTaskConfiguration(task), this.getContext());
+        } catch (error) {
+            this.logger.error(`Error launching task '${taskName}': ${error}`);
+            this.messageService.error(`Error launching task '${taskName}': ${error}`);
+            return;
+        }
+
+        this.logger.debug(`Task created. task id: ${taskInfo.taskId}, OS ProcessId: ${taskInfo.osProcessId} `);
+
+        // open terminal widget if the task is based on a terminal process:
+        if (taskInfo.terminalId !== undefined) {
+            this.attachTerminal(taskInfo.terminalId, taskInfo.taskId);
+        }
+    }
+
+    public async attach(terminalId: number, taskId: number): Promise<void> {
+        this.attachTerminal(terminalId, taskId);
+    }
+
+    /**
+     * Perform some adjustments to the task launch configuration, before sending
+     * it to the backend to be executed. We can make sure that parameters that
+     * are optional to the user but required by the server will be defined, with
+     * sane default values.
+     */
+    protected massageTaskConfiguration(task: TaskOptions): TaskOptions {
+        if (task.cwd) {
+            if (this.workspaceRootUri) {
+                task.cwd = task.cwd.replace(/\$workspace/gi, this.workspaceRootUri);
+            }
+        } else if (this.workspaceRootUri) {
+            // if "cwd' is not set in task launch config. Let's use the workspace root
+            // as default
+            task.cwd = this.workspaceRootUri;
+        }
+
+        if (!task.processType) {
+            task.processType = 'terminal';
+        }
+        return task;
+    }
+
+    protected async attachTerminal(terminalId: number, taskId: number): Promise<void> {
+        // create terminal widget to display task's execution output
+        const widget = <TerminalWidget>await this.widgetManager.getOrCreateWidget(
+            TERMINAL_WIDGET_FACTORY_ID,
+            <TerminalWidgetFactoryOptions>{
+                created: new Date().toString(),
+                id: 'task-' + taskId,
+                caption: `Task #${taskId}`,
+                label: `Task #${taskId}`,
+                destroyTermOnClose: true
+            });
+        widget.start(terminalId);
+    }
+
+    private isEventForThisClient(context: string | undefined): boolean {
```

Comment:
protected

Reply:
done. BTW, I noticed in general that we tend not to explicitly mark methods as "public". I was doing it in this class, so I removed these.

Reply:
OK yes that's still a gray zone to me.